internal/tracer/middleware.go

```diff
@@ -39,12 +40,24 @@ func TracingRoundTripper(transport http.RoundTripper, collector Collector) http.
 		resp, err := transport.RoundTrip(req)
 		if err != nil {
 			builder.add(&ResponseError{Err: err})
-			cancel()
 			return nil, err
 		}
-		builder.add(&ResponseStart{Response: resp})
+
+		defer resp.Body.Close()
+		// Read the response body in full and then replace it with a new in-memory reader
+		// so that we can read the contents multiple times in any trace or middleware.
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		resp.Body = io.NopCloser(bytes.NewReader(body))
+		builder.add(&ResponseStart{
+			Response: resp,
+		})
+
 		respClone := *resp
-		respClone.Body = newReader(resp.Header, resp.Body, false, builder, cancel)
+		respClone.Body = newReader(resp.Header, io.NopCloser(bytes.NewReader(body)), false, builder, cancel)
+
```

Comment:
I don't love this, but not sure if there's a better way. I considered two other approaches:

* Storing the `body` data on the `ResponseStart` event so that it could be accessed by the `Complete` function override. The response body is needed there for reading unary details. 
* Creating a reusable reader that could be read 1-to-many times, but that kind of felt like overkill. 

We basically need the body data twice -- once inside the `Complete` function and once inside the `trace.Reader`. So, I opted for reading it once from the network here and then using an in-memory reader in both places. Def open to a better way though.

Reply:
Oh, also, the `io.ReadAll` call seems to hang for full-duplex streams that return a result. It works fine for every other type (including half-duplex AND full-duplex that do not return a response). Not sure if you're aware of anything that would be causing that.

Reply:
Yep, you can't do this in a round-tripper. See the comment I just added above. You need middleware to replace the `Response.Body` with an `io.ReadCloser` that will capture the data as it is read.